users/mkillewald/mkillewald.c

```diff
@@ -0,0 +1,283 @@
+            set_mac_base_rgb();
+            break;
+        case WIN_BASE:
+            set_win_base_rgb();
+            break;
+        default:
+            break;
+    }
+    return default_layer_state_set_keymap(state);
+}
+
+void set_mac_base_rgb(void) {
+    // load mac base settings
+    rgb_matrix_enable_noeeprom();
+    rgb_matrix_reload_from_eeprom();
+
+    // check disable
+    if (!user_config_get_enable_mac_base()) {
+        rgb_matrix_disable_noeeprom();
+    }
+}
+
+void set_win_base_rgb(void) {
+    // load win base settings
+    rgb_matrix_enable_noeeprom();
+    rgb_matrix_mode_noeeprom(user_config_get_mode_win_base());
+    rgb_matrix_set_speed_noeeprom(user_config_get_spd_win_base());
+    rgb_matrix_sethsv_noeeprom(user_config_get_hsv_win_base().h,
+                               user_config_get_hsv_win_base().s,
+                               user_config_get_hsv_win_base().v);
+
+    // check disable
+    if (!user_config_get_enable_win_base()) {
+        rgb_matrix_disable_noeeprom();
+    }
+}
+
+__attribute__ ((weak))
+void matrix_init_keymap(void) {
+  return;
+}
+
+void matrix_init_user(void) {
+#ifdef RGB_MATRIX_ENABLE
+    rgb_matrix_init_user();
+#endif
+
+    matrix_init_keymap();
+}
+
+__attribute__ ((weak))
+void keyboard_post_init_keymap(void) {
+  return;
+}
+
+void keyboard_post_init_user(void) {
+    user_config_read_eeprom();
+
+#ifdef AUTOCORRECT_OFF_AT_STARTUP
+    // toggle autocorrect off at startup
+    if (autocorrect_is_enabled()) { autocorrect_toggle(); }
+#endif
+
+    keyboard_post_init_keymap();
+}
+
+void housekeeping_task_mkillewald(void) {
+    switch(bootloader_state) {
+        case BOOTLOADER_DO:
+            // bootloader was pressed two frames ago. RGB should now be off,
+            // so we can call the bootloader.
+            reset_keyboard();
+            break;
+        case BOOTLOADER_WAIT:
+            // bootloader was pressed on previous frame, we wait this frame and
+            // set flag to do bootloader at end of next frame. For some reason, my
+            // Q2 needed this extra wait frame.
+            bootloader_state = BOOTLOADER_DO;
+            break;
+        case BOOTLOADER_PRESSED:
+            // User pressed bootloader keycode and RGB was disabled earlier in this
+            // frame. However RGB changes wont take place immediately, so we set a
+            // flag here which will be caught at end of the next frame.
+            bootloader_state = BOOTLOADER_WAIT;
+            break;
+        default:
+            break;
+    }
+}
+
+bool process_record_mkillewald(uint16_t keycode, keyrecord_t *record) {
+    switch (keycode) {
+        case QK_BOOT:
+            // We want to turn off LEDs before calling bootloader, so here
+            // we call rgb_matrix_disable_noeeprom() and set a flag because
+            // the LEDs won't be updated until the next frame.
+            if (record->event.pressed) {
+                rgb_matrix_disable_noeeprom();
```

Comment:
you can actually use `void shutdown_user(void)`, as this function gets called by most of the reset functions (like `QK_BOOT`).

Reply:
Ah, yeah, it can be a bit tricky. This will work to shut off all the rgb: 
```c
void shutdown_user(void) {
    // to disable RGB when entering DFU mode from QK_BOOT keycode
    rgb_matrix_set_color_all(0x00, 0x00, 0x00);
    rgb_matrix_update_pwm_buffers();
}
```

Reply:
Ahh, that worked. Thank you!